src/platform/forms-system/src/js/fields/FileField.jsx

```diff
@@ -29,25 +39,96 @@ export default class FileField extends React.Component {
     }
   }
 
-  onAddFile = (event, index = null) => {
+  componentDidMount() {
+    // The File object is not preserved in the save-in-progress data
+    // We need to remove these entries; an empty `file` is included in the
+    // entry, but if API File Object still exists (within the same session), we
+    // can't use Object.keys() on it because it returns an empty array
+    const formData = (this.props.formData || []).filter(
+      // keep - file may not exist (already uploaded)
+      // keep - file may contain File object; ensure name isn't empty
+      // remove - file may be an empty object
+      data => !data.file || (data.file?.name || '') !== '',
+    );
+    if (formData.length !== (this.props.formData || []).length) {
+      this.props.onChange(formData);
+    }
+  }
+
+  focusAddAnotherButton = () => {
+    // focus on span pseudo-button, not the label
+    focusElement(`#${this.props.idSchema.$id}_add_label span`);
+  };
+
+  onSubmitPassword = (file, index, password) => {
+    if (file && password) {
+      this.onAddFile({ target: { files: [file] } }, index, password);
+    }
+  };
+
+  isFileEncrypted = async file =>
+    checkForEncryptedPdf(file)
+      .then(isEncrypted => isEncrypted)
+      // This _should_ only happen if a file is deleted after the user selects
+      // it for upload
+      .catch(() => false);
+
+  /**
+   * Add file to list and upload
+   * @param {Event} event - DOM File upload event
+   * @param {number} index - uploaded file index, if already uploaded
+   * @param {string} password - encrypted PDF password, only defined by
+   *   `onSubmitPassword` function
+   * @listens
+   */
+  onAddFile = async (event, index = null, password) => {
     if (event.target.files && event.target.files.length) {
+      const currentFile = event.target.files[0];
       const files = this.props.formData || [];
+      const { requestLockedPdfPassword, uiSchema } = this.props;
+      const { getEncryptedPassword } = uiSchema['ui:options'];
+
       let idx = index;
       if (idx === null) {
         idx = files.length === 0 ? 0 : files.length;
       }
+
+      // Check if the file is an encrypted PDF
+      if (
+        requestLockedPdfPassword && // feature flag
+        getEncryptedPassword && // uiSchema flag
+        currentFile.name?.endsWith('pdf') &&
+        !password
+      ) {
+        const needsPassword = await this.isFileEncrypted(currentFile);
+        if (needsPassword) {
+          files[idx] = {
+            file: currentFile,
+            name: currentFile.name,
+            isEncrypted: true,
+          };
+          this.props.onChange(files);
+          // wait for user to enter a password before uploading
+          return;
```

Comment:
It may be useful to have a unit test to check that an `onChange` is called in this situation but `uploadFile` is _not_ called.